### ISSUE TITLE
inductor: there has an error when pow's exp is not integer for symbolic shape

### ISSUE THREAD

**Issue by XiaobingSuper** [pytorch/pytorch#97877]
### 🐛 Describe the bug

For dynamic shape case for timm ```convit_base```, there has an error:
```
  File "/home/xiaobing/pytorch-offical/torch/_inductor/codegen/wrapper.py", line 110, in codegen
    line = self.wrapper.make_buffer_allocation(self.node)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/codegen/wrapper.py", line 572, in make_buffer_allocation
    f"{self.codegen_shape_tuple(shape)}, "
  File "/home/xiaobing/pytorch-offical/torch/_inductor/codegen/wrapper.py", line 434, in codegen_shape_tuple
    parts = list(map(self.codegen_sizevar, shape))
  File "/home/xiaobing/pytorch-offical/torch/_inductor/codegen/wrapper.py", line 431, in codegen_sizevar
    return pexpr(V.graph.sizevars.simplify(x))
  File "/home/xiaobing/miniconda3/envs/pytorch_inductor/lib/python3.9/site-packages/sympy/printing/printer.py", line 292, in doprint
    return self._str(self._print(expr))
  File "/home/xiaobing/miniconda3/envs/pytorch_inductor/lib/python3.9/site-packages/sympy/printing/printer.py", line 331, in _print
    return printmethod(expr, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/codegen/common.py", line 117, in _print_floor
    return f"math.floor({self._print(expr.args[0])})"
  File "/home/xiaobing/miniconda3/envs/pytorch_inductor/lib/python3.9/site-packages/sympy/printing/printer.py", line 331, in _print
    return printmethod(expr, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/codegen/common.py", line 77, in _print_Pow
    assert exp.is_integer
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
AssertionError:


You can suppress this exception and fall back to eager by setting:
    torch._dynamo.config.suppress_errors = True
```

there a smaller case can reproduce this error:

```
import torch
import torch._dynamo
import torch._dynamo.config as config


config.dynamic_shapes=True

class Model(torch.nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        B, N, C = x.shape
        return self.get_rel_indices(N)

    def get_rel_indices(self, num_patches: int) -> torch.Tensor:
        img_size = int(num_patches ** .5)
        #rel_indices = torch.zeros(1, num_patches, num_patches, 3)
        ind = torch.arange(img_size)
        return ind

model = Model().eval()
opt_model = torch._dynamo.optimize('inductor')(model)

batch_size = 4
x = torch.randn(8, 196, 768)
ref = model(x)
with torch.no_grad():
    for i in range(3):
        out = opt_model(x)
```

seems that the symbolic pow requires the exp should be an integer.

### Versions

Collecting environment information...
PyTorch version: 2.1.0a0+git33dfded
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.31

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.13.0-30-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] clip-anytorch==2.5.2
[pip3] CoCa-pytorch==0.0.7
[pip3] dalle2-pytorch==1.10.5
[pip3] ema-pytorch==0.2.2
[pip3] functorch==1.14.0a0+408bcf1
[pip3] mypy==0.960
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.1
[pip3] pytorch-transformers==1.2.0
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.2.1
[pip3] torch==2.1.0a0+gitdc85921
[pip3] torch-fidelity==0.3.0
[pip3] torch-struct==0.5
[pip3] torchaudio==2.0.0a0+a8f4e97
[pip3] torchdata==0.7.0a0+f1283eb
[pip3] torchmetrics==0.11.4
[pip3] torchrec-nightly==2023.3.23
[pip3] torchtext==0.15.0a0+46e7eef
[pip3] torchvision==0.15.0a0+98c5815
[pip3] vector-quantize-pytorch==1.1.2
[conda] bert-pytorch              0.0.1a4                   dev_0    <develop>
[conda] clip-anytorch             2.5.2                    pypi_0    pypi
[conda] coca-pytorch              0.0.7                    pypi_0    pypi
[conda] dalle2-pytorch            1.10.5                   pypi_0    pypi
[conda] ema-pytorch               0.2.2                    pypi_0    pypi
[conda] functorch                 1.14.0a0+408bcf1          pypi_0    pypi
[conda] mkl                       2023.0.0         h6d00ec8_25399
[conda] mkl-include               2023.0.0         h06a4308_25399
[conda] numpy                     1.23.1                   pypi_0    pypi
[conda] pytorch-transformers      1.2.0                    pypi_0    pypi
[conda] pytorch-warmup            0.1.1                    pypi_0    pypi
[conda] rotary-embedding-torch    0.2.1                    pypi_0    pypi
[conda] torch                     2.1.0a0+gitdc85921           dev_0    <develop>
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torch-struct              0.5                      pypi_0    pypi
[conda] torchaudio                2.0.0a0+a8f4e97          pypi_0    pypi
[conda] torchdata                 0.6.0                    pypi_0    pypi
[conda] torchmetrics              0.11.4                   pypi_0    pypi
[conda] torchrec-nightly          2023.3.23                pypi_0    pypi
[conda] torchtext                 0.15.0a0+46e7eef          pypi_0    pypi
[conda] torchvision               0.15.0a0+98c5815           dev_0    <develop>
[conda] vector-quantize-pytorch   1.1.2                    pypi_0    pypi

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

**Comment by XiaobingSuper** [comment]
@ezyang, from the comments at https://github.com/pytorch/pytorch/blob/4ae4c6f68ab64ddf04026419b39d6095fe2f0513/torch/_inductor/codegen/common.py#L68-L76,

the suggestion is that ```make upstream retranslate the Sympy expression into Tensor expressions earlier```, how to do the retranslate? Thanks!

**Comment by ezyang** [comment]
Unfortunately I don't have a clear picture of how to do this. https://docs.google.com/document/d/1INSCdYu1PxXcr43HrD82OudeEuS-qxQe1yZmLg2wy6A/edit#heading=h.aytveihghlgn says directionally what the plan is, but the concrete nuts and bolts need to be figure dout

**Comment by jgong5** [comment]
According to this test case, does it mean the assumption that "You don't typically expect to have to do floating point computation including exponents in sizevar compute" does not hold? I'm not sure how hard it is to retranslate things to Tensor from the upstream but would it be simpler to handle floating point exponent in this print function?

**Comment by ezyang** [comment]
Reading over the test case carefully, you're correct. I guess let's add sqrt support specifically directly here. cc @ngimel
